Shopify/hydrogen
# Add JSDocs to generated projects

### WHY are these changes introduced?

We already have GraphQL Codegen for TypeScript but JS projects are not getting any benefits from it. This adds JSDocs to the generated JS projects.

### WHAT is this pull request doing?

This uses [`ts-morph`](https://github.com/dsherret/ts-morph) to transpile TS => JS and also generate JSDocs at the same time from the TS types.

First, it adds typedefs for each type at the bottom:

<img width="915" alt="image" src="https://github.com/Shopify/hydrogen/assets/1634092/397f821f-98b8-4056-a031-717fcecda897">

Then, it annotates each function using the types defined:

<img width="695" alt="image" src="https://github.com/Shopify/hydrogen/assets/1634092/b80fcfbf-a00b-4d87-bac9-004ad8eecd97">

### HOW to test your changes?

In the `templates` folder, run:

```bash
h2 init --path jsdocs --language js --force --styling none --mock-shop --no-shortcut --markets domains --no-install-deps --routes && cd jsdocs && ln -s ../../node_modules node_modules
```

Check that the generated project has JSDocs and works with Codegen. You might need to `>TypeScript: Restart TS Server` in VSCode at times.

#### Post-merge steps



#### Checklist
